pkg/datasource/datasource.go

```diff
@@ -2,119 +2,107 @@ package datasource
-
-	ds.QueryMux = mux
+func NewDatasource(ctx context.Context, xrayClientFactory XrayClientFactory, settings *awsds.AWSDatasourceSettings) *Datasource {
+	ds := &Datasource{xrayClientFactory: xrayClientFactory, Settings: *settings}
 
+	// resource handler
 	resMux := http.NewServeMux()
 	resMux.HandleFunc("/groups", ds.getGroups)
 	resMux.HandleFunc("/regions", ds.GetRegions)
 	resMux.HandleFunc("/accounts", ds.GetAccounts)
-
 	ds.ResourceMux = httpadapter.New(resMux)
+
+	ds.authSettings, _ = awsds.ReadAuthSettingsFromContext(ctx)
+	ds.sessions = awsds.NewSessionCache()
 	return ds
 }
 
+func (ds *Datasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	res := backend.NewQueryDataResponse()
+	for _, query := range req.Queries {
+		var currentRes backend.DataResponse
+		switch query.QueryType {
+		case QueryGetTrace:
+			currentRes = ds.getSingleTrace(ctx, query, &req.PluginContext)
+		case QueryGetTraceSummaries:
+			currentRes = ds.getTraceSummariesForSingleQuery(ctx, query, &req.PluginContext)
+		case QueryGetTimeSeriesServiceStatistics:
+			currentRes = ds.getTimeSeriesServiceStatisticsForSingleQuery(ctx, query, &req.PluginContext)
+		case QueryGetAnalyticsRootCauseResponseTimeService,
+			QueryGetAnalyticsRootCauseResponseTimePath,
+			QueryGetAnalyticsRootCauseErrorService,
+			QueryGetAnalyticsRootCauseErrorPath,
+			QueryGetAnalyticsRootCauseErrorMessage,
+			QueryGetAnalyticsRootCauseFaultService,
+			QueryGetAnalyticsRootCauseFaultPath,
+			QueryGetAnalyticsRootCauseFaultMessage,
+			QueryGetAnalyticsUser,
+			QueryGetAnalyticsUrl,
+			QueryGetAnalyticsStatusCode:
+			currentRes = ds.getSingleAnalyticsQueryResult(ctx, query, &req.PluginContext)
+		case QueryGetInsights:
+			currentRes = ds.getSingleInsight(ctx, query, &req.PluginContext)
+		case QueryGetServiceMap:
+			currentRes = ds.getSingleServiceMap(ctx, query, &req.PluginContext)
+		default:
+			res.Responses[query.RefID] = backend.DataResponse{
+				Error: fmt.Errorf("unknown query type: %s", query.QueryType),
+			}
+			continue
+		}
+		res.Responses[query.RefID] = currentRes
+	}
+	return res, nil
+}
+
+func (ds Datasource) CallResource(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
+	return ds.ResourceMux.CallResource(ctx, req, sender)
+}
+
 type RequestSettings struct {
 	Region string
 }
 
-func getXrayClient(ctx context.Context, pluginContext *backend.PluginContext, requestSettings RequestSettings) (XrayClient, error) {
+func (ds Datasource) getClient(ctx context.Context, pluginContext *backend.PluginContext, requestSettings RequestSettings) (XrayClient, error) {
+	return ds.xrayClientFactory(ctx, pluginContext, requestSettings, ds.authSettings, ds.sessions)
+}
+
+func getXrayClient(ctx context.Context, pluginContext *backend.PluginContext, requestSettings RequestSettings, authSettings *awsds.AuthSettings, sessions *awsds.SessionCache) (XrayClient, error) {
```

Comment:
nit: Here and elsewhere `backend.PluginContext`, `backend.DataSourceInstanceSettings`, `awsds.AuthSettings` and `awsds.AWSDatasourceSettings` don't need to be passed around or stored as pointers. This is a bit of a hobby horse of mine - I think using pointers where they're not necessary makes code harder to understand - so ignore me if you like 😅 but I pushed a branch [here](https://github.com/grafana/x-ray-datasource/compare/iwysiu/229..njvrzm/streamline_pointers) if you want to cherry-pick it. Entirely up to you.

Reply:
cherry-picked!